File: Python/Leetcode Medium/621TaskScheduler.py
#https://leetcode.com/problems/task-scheduler/description/
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        dictionary = {}
        for task in tasks: 
            if task in dictionary: dictionary[task] += 1
            else: dictionary[task] = 1
        heap = [] #Number of tasks still there, Interval Waiting
        waitingQueue = deque([])
        for task in dictionary: heapq.heappush(heap, [-1 * dictionary[task], 0]) 
        clock = 0
        while len(heap) > 0 or len(waitingQueue) > 0:
            while len(waitingQueue) > 0 and waitingQueue[0][1] <= clock: heapq.heappush(heap, waitingQueue.popleft())
            if len(heap) == 0:
                clock += 1
                continue
            task = heapq.heappop(heap)
            if task[1] <= clock:
                count = task[0] + 1
                if count != 0: waitingQueue.append([count, clock+n+1])
            clock += 1
        return clock
    
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        dictionary = {}
        for task in tasks:
            if task in dictionary: dictionary[task] += 1
            else: dictionary[task] = 1
        heap = []
        for task in dictionary: heapq.heappush(heap, [0, task, dictionary[task]])
        clock = 0
        while len(heap) > 0:
            value = heapq.heappop(heap)
            if value[0] <= clock:
                num = value[2] - 1
                if num != 0: heapq.heappush(heap, [value[0]+n+1, task, num])
            else:
                heapq.heappush(heap, value)
            clock += 1
        return clock

File: C++/Leetcode Medium/621TaskScheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        //Count freq of each task
        unordered_map<char, int> freq;
        for (char task : tasks) ++freq[task];

        //Initialize max heap for ready tasks and queue for waiting tasks
        priority_queue<tuple<int, char>> ready; //freq, task
        for (auto [task, frequency] : freq) ready.push({frequency, task});
        deque<tuple<int, int, char>> queue; //time, freq, task

        //Core Algorithm
        int res = 0;
        while (!ready.empty() || !queue.empty()) {
            //Move what's been waiting to ready if ready
            while (!queue.empty() && get<0>(queue.front()) <= res) {
                ready.push({get<1>(queue.front()), get<2>(queue.front())});
                queue.pop_front();
            }

            //If ready is empty, update res so that it's atleast the next value of time in the front tuple of waiting
            if (ready.empty()) {res = get<0>(queue.front()); continue;}
            
            //Get best candidate in ready (task with highest freq)
            auto [currFreq, currTask] = ready.top();
            ready.pop();

            //Add to waiting if num of tasks is > 1
            if (currFreq-1 > 0) queue.push_back({res + n + 1, currFreq-1, currTask});
            ++res;
        }

        return res;
    }
};
//Count the frequency of each task
//max heap (tasks that are ready) 
//queue (tasks that are waiting that will eventually go back to the heap)